### PR TITLE
Improve signature checks in config_api

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
The intended purpose of storing signer Pubkeys in Config data is to transfer ownership from the Config account keypair to another keypair(s). However, if the stored keypairs are not checked on an account update, any signing keypairs can overwrite data with properly constructed instruction data.

#### Summary of Changes
- On initialization, the Config account keypair is required to be a signer
- On updates, the Config account keypair is only required to be a signer if specified in Config account data
- For a successful update, either the Config account keypair must be a signer, or the set of signers must equal the set of required signers previously specified in Config account data

@rob-solana , I think this addresses your question on #4980 , but I am not totally sure I've understood the question properly.

### CODE REVIEW

**Comment by mvines** [comment]
I think this scheme gets us to the same place without the extra boolean in the Account data:
1. On initialization, the Config account keypair is the only required signer
2. On updates,
      a. the Config account keypair is the only required signer if no signer keys are specified in Config account data
     b. Otherwise the signers keys are specified in the Config account (which may include the Config account pubkey)

**Comment by CriesofCarrots** [comment]
> I think this scheme gets us to the same place without the extra boolean in the Account data:
> 
>     1. On initialization, the Config account keypair is the only required signer
> 
>     2. On updates,
>        a. the Config account keypair is the only required signer if no signer keys are specified in Config account data
>        b. Otherwise the signers keys are specified in the Config account (which may include the Config account pubkey)

Yes, it does, but requires an entire extra `(Pubkey, bool)` in Account data for "Config + other signers" updates.
Guess I don't feel that strongly about it, though.

**Comment by mvines** [comment]
Yeah it treats the Config account pubkey like the rest, I guess I prefer not having that Boolean special myself.

**Comment by CriesofCarrots** [comment]
@mvines , more better?

File: programs/config_api/src/config_processor.rs [hunk 1]
```diff
@@ -16,6 +16,27 @@ pub fn process_instruction(
...
+        InstructionError::InvalidAccountData
+    })?;
+    let current_signer_keys: Vec<Pubkey> = current_data
+        .keys
+        .iter()
+        .filter(|(_, is_signer)| *is_signer)
+        .map(|(pubkey, _)| *pubkey)
+        .collect();
+
+    if current_data.keys.is_empty() || current_signer_keys.is_empty() {
```
Comment (mvines) [inline programs/config_api/src/config_processor.rs]: I think we can just test for `current_signer_keys.is_empty()` here

**Pull request merged**